### ISSUE TITLE
[Editor] Migrate all the editor internal events to the new event system

### ISSUE THREAD

**Issue by shaddockh** [AtomicGameEngine/AtomicGameEngine#1424]
Convert all the events used inside the editor to use the new Atomic Native and Atomic Script events.

**Comment by shaddockh** [comment]
@JoshEngebretson @mattbenic As I'm going through this, I'm looking at all of the events that are defined inside ```EditorEvents.ts``` at https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Script/AtomicEditor/editor/EditorEvents.ts 

What I would like to do is migrate these events to the native c++ and update the references.  I have two reasons for this.
1) It standardizes all the editor events and will auto-generate the event handlers that can be used on build
2) As native events, they will be accessible to code w/o having to explicitly import ```EditorEvents.ts```.  Most specifically, this will allow extensions to easily subscribe and emit these events w/o having to make a copy of that file.

My questions are does this sound like a good idea and @mattbenic, will this affect what you are doing with extensions?

**Comment by JoshEngebretson** [comment]
I think this is a good idea for core editor events, I think we can also use ScriptEvents for anything anything "script-y" editor side, which I am not entirely sure what I mean by that, though will know it when we see it.  Basically, working on the editor itself shouldn't require dropping into C++ at any kind of regular intervals.

I also think this is a good update, as getting the core editor events in C++ for event generation will also work for C# if we go with a managed editor approach, one thing that could be really awesome there, is being able to extend the editor in both C# and TypeScript, this has the potential of getting the AtomicNETService in the editor process too, which would simplify some things.  Just brainstorming, would love to hear thoughts :)

**Comment by shaddockh** [comment]
Interesting idea! :) . If more functionality is accessible through events, then the C# and TS could communicate with each other strictly through the event system.  For instance, there could be an event ```AddMenuItemEvent``` that C# could send and the TS could receive and create the menu item for.  Then the C# could subscribe to a menu click event and then perform some kind of action.  Basically some kind of IPC system utilizing the events passing through the native layer.

Could have possibilities!

Anyway, I'll move forward in migrating the core editor events to native.

**Comment by shaddockh** [comment]
@JoshEngebretson I think I ended up falling down a rabbit hole! :)

Anyway, I'm currently going through and looking at adding type safety to the ```sendEvent``` calls.

The current signature of send event is: 
```sendEvent(string, object?)``` 
which provides no type checking whatsoever.

I've just spent the last couple of nights adding some type safety and have come up with the following overload:
```sendEvent<T extends Atomic.EventMetadata>(name:Atomic.EventType, data: T);```

Which results in usage that looks something like this:
```
   object.sendEvent<Editor.SceneEditComponentCopyEvent>(Editor.SceneEditComponentCopyEventType, { component: this });
```
This ensures that the string being passed in for the event type is a valid Atomic.EventType and that the data conforms to the right structure.  But I'm wondering if I should instead follow the pattern that's being used for the ```subscribeToEvent``` instead.

That would be:
```sendEvent<T extends Atomic.EventData>( eventData: T );```

Which would end up being called like:
```
object.sendEvent(Editor.SceneEditComponentCopyEventData({component: this}));
```
(TS will implicitly figure out the T so you don't have to explicitly cast it, which reduces keystrokes)

This could be accomplished with the same magic that is happening for the events and functions could be dynamically added to the appropriate namespace to be called.  Once called, the passed in data object would be augmented with the correct event type string that could then be extracted inside the ```sendEvent``` method.

I think this may be more work, but will ultimately make things much cleaner.   And it ensures that the proper event type is associated with the data and the data is type checked. Thoughts?
